### PR TITLE
Fix #746, #760, #946: enhance the My Feeds page

### LINKED ISSUE #746: My Feeds page: implement button functionalities
Presently, the implied/intended functionalities of the ['+' and '?' buttons on the My Feeds page](https://github.com/Seneca-CDOT/telescope/pull/735) have not been implemented. These functionalities should be added during a future release.

<hr>

- Intended behaviour of the '?' button: 
    - As per [our design](https://github.com/Seneca-CDOT/telescope/issues/725#issuecomment-586726492), when this button is clicked, some type of [inline help box](https://ui-patterns.com/patterns/InlineHelpBox) should be rendered. This help box should contain text that briefly informs the user what they need to input into a `url` text field.
- Intended behaviour of the '+' button:
  - If clicked when the user has no feeds saved (i.e. has never input anything into a `url` text field), this should simply focus the user's input into the solitary `url` text field rendered to the view (i.e. _not_ create another field).
  - Otherwise, clicking this button should render a new empty `url` text field (and associated icon) below all preexisting fields.
    - **Design decision:** as there need only be one '?' button rendered to the view at one time, the optimal location of this button is yet to be determined. For example, this button could have a fixed alignment alongside the first blog feed URL input field (whether empty or not), or the button's vertical position could shift downward to align with the newly-added, empty input field.

### LINKED ISSUE #760: Restrict access to My Feeds to logged-in users
Once https://github.com/Seneca-CDOT/telescope/pull/748 is landed, anyone on the Internet will be able to add feeds to Telescope. Access to this component should be restricted to logged-in users, ensuring that non-users cannot add unwanted feeds (e.g. spam) to our system.

### LINKED ISSUE #946: My Feeds page: connect feed deletion button to backend functionality
**What would you like to be added**:

The implementation of the feed deletion buttons on the My Feeds page (introduced by #933) must be completed, i.e. by invoking [the to-be backend logic for feed deletion](https://github.com/Seneca-CDOT/telescope/pull/919) that is being developed by @c3ho and @Grommers00.

**Why would you like this to be added**:

This is one step towards ensuring that the My Feeds page is capable of facilitating a vital administration functionality.

### PR DESCRIPTION
**Pull request by Silvyre**
## Issue This PR Addresses

Fixes #746
Fixes #760
Fixes #946

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [x] **UI**: Change which improves UI

## Description

Fixes #746 by finalizing the implementation of buttons on the My Feeds page:
- Clicking on the `?` button of the My Feeds page now opens a [popover](https://material-ui.com/api/popover/) containing a helpful message.
   - Uses [`material-ui-popup-state`](https://www.npmjs.com/package/material-ui-popup-state) (simplifies popover state management)
- Introduces form handling/validation via [`react-material-ui-form-validator`](https://www.npmjs.com/package/react-material-ui-form-validator).
   - (One of the custom-defined validation rules uses the `valid-url` module, which was previously added to Telescope for use in our backend.)
- Calling `addFeed()` now causes a status message to be rendered to the page, indicating whether its fetch request to add a feed succeeded or failed.
- When the page is loaded, each of the current user's owned feeds are fetched and rendered to the page within separate (disabled) text boxes.

Fixes #760 by preventing access to the page by users who are not logged in.
- See: https://github.com/Seneca-CDOT/telescope/issues/760#issuecomment-610695732

Fixes #946 by adding feed deletion buttons alongside each existing feed.
   - **_Update: these now send a `DELETE /feeds/:id` request when clicked._**

_How to review:_
1. Run `docker-compose up` (ensure that `API_URL=http://localhost:3000` in `.env`)
1. Run `npm run clean; npm run build` to build the Gatsby frontend
1. Log into Telescope (http://localhost:3000/auth/login, user1/user1pass)
1. Navigate to the My Feeds page (http://localhost:3000/myfeeds)
1. Fill out the two form controls and click the '+' button to add a new feed. Doing so should cause the new feed to (after a few seconds) appear as a new set of form controls.
1. Clicking on the '?' button should pop up an informative message to assist with the previous step.
1. Wait several seconds after the page refreshes: the new feed should appear within an additional set of (disabled) form controls.
1. Clicking on the 'delete' icon (next to the set of disabled controls) should spawn a dialog window confirming the action. Confirming this dialog will send a `DELETE /feeds/:id` request (which won't result in anything happening until the corresponding backend functionality is merged and this PR is rebased over it).

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by Silvyre** [comment]
Update as of https://github.com/Seneca-CDOT/telescope/pull/933/commits/50bef0e986fc634ada96d801c4cbbdbab7be5def:

> - When the page is loaded, each of the current user's owned feeds should be fetched and rendered to the page within separate <TextValidator> controls.

This is mostly done; user-owned feeds are fetched and used to construct additional sets of controls. However, these constructed are not yet functional.

**Comment by Silvyre** [comment]
Alright, this is about ready for review! I updated the PR's description with the steps I recommend taking to review it.

File: src/frontend/src/pages/myfeeds.js [hunk 1]
```diff
@@ -15,94 +34,312 @@ const useStyles = makeStyles((theme) => ({
...
+  const newFeedAuthorRef = React.createRef();
+  const newFeedUrlRef = React.createRef();
+
+  async function hashUserFeeds() {
+    const [user, feedItems] = await Promise.all([
+      fetch(`${telescopeUrl}/user/info`).then((res) => res.json()),
+      fetch(`${telescopeUrl}/feeds`).then((res) => res.json()),
+    ]);
+    const allFeeds = await Promise.all(
+      feedItems.map((item) => fetch(`${telescopeUrl}${item.url}`))
```
Comment (Silvyre) [inline src/frontend/src/pages/myfeeds.js:157]: This approach admittedly results in about 700 fetch requests and probably should be refactored. Up for any suggestions.

**Comment by cindyledev** [comment]
This is super cool! I really like what you did with the popup module
Annotation 2020-04-06 083653

I see you're still working on the deletion functionality so I'll leave you to that. Sorry I can't really help in this area, my backend expertise is limited :(

File: src/frontend/src/pages/myfeeds.js [hunk 2]
```diff
@@ -1,7 +1,26 @@
...
+  Grid,
+  IconButton,
+  Popover,
+  Typography,
+} from '@material-ui/core';
+import { AccountCircle, AddCircle, Delete, HelpOutline, RssFeed } from '@material-ui/icons';
 import { makeStyles } from '@material-ui/core/styles';
+import { ValidatorForm, TextValidator } from 'react-material-ui-form-validator';
+import PopupState, { bindTrigger, bindPopover } from 'material-ui-popup-state';
+// eslint-disable-next-line import/no-extraneous-dependencies
```
Comment (humphd) [inline src/frontend/src/pages/myfeeds.js:22]: Why is this necessary?

File: src/frontend/src/pages/myfeeds.js [hunk 3]
```diff
@@ -15,94 +34,312 @@ const useStyles = makeStyles((theme) => ({
...
+    setOpen(true);
+  };
+
+  const handleClose = () => {
+    setOpen(false);
+  };
+
+  const removeFeed = () => {
+    console.log(`Removing feed hosted at URL ${feed.url}`);
+    // TODO
```
Comment (humphd) [inline src/frontend/src/pages/myfeeds.js:93]: File a follow-up issue for this, and put the URL to the issue here.

Comment (humphd) [reply src/frontend/src/pages/myfeeds.js:93]: Also, cc @c3ho and @Grommers00, who are working on backend bits for this.

File: src/frontend/src/pages/myfeeds.js [hunk 4]
```diff
@@ -15,94 +34,312 @@ const useStyles = makeStyles((theme) => ({
...
+    window.location.reload(false);
```
Comment (humphd) [inline src/frontend/src/pages/myfeeds.js:94]: Why?

File: src/frontend/src/pages/myfeeds.js [hunk 5]
```diff
@@ -15,94 +34,312 @@ const useStyles = makeStyles((theme) => ({
...
+      <IconButton classes={{ root: classes.button }} onClick={handleClickOpen}>
+        <Delete color="secondary" />
+      </IconButton>
+      <Dialog
+        open={open}
+        onClose={handleClose}
+        aria-labelledby="alert-dialog-title"
+        aria-describedby="alert-dialog-description"
+      >
+        <DialogTitle id="alert-dialog-title">{`Remove feed hosted at ${feed.url}?`}</DialogTitle>
```
Comment (humphd) [inline src/frontend/src/pages/myfeeds.js:108]: You might want to think about how you'll deal with really long URLs here.  I checked, and the longest one we have in the wiki is 87 characters (average is 51 characters).

File: src/frontend/src/pages/myfeeds.js [hunk 6]
```diff
@@ -15,94 +34,312 @@ const useStyles = makeStyles((theme) => ({
...
 
-  function handleAuthorChange(author) {
-    setFeedAuthor(author);
-  }
+  const newFeedAuthorRef = React.createRef();
+  const newFeedUrlRef = React.createRef();
+
+  async function hashUserFeeds() {
+    const [user, feedItems] = await Promise.all([
+      fetch(`${telescopeUrl}/user/info`).then((res) => res.json()),
```
Comment (humphd) [inline src/frontend/src/pages/myfeeds.js:153]: All of this can/will fail at some point, we need proper error handling.  You can chat with @c3ho or @raygervais, who have recently done code like this with error handling.

Comment (humphd) [reply src/frontend/src/pages/myfeeds.js:153]: I also think we should probably move the call to get user info to the top level of this page, and already know who you are when you access the content.  That's something for #760.

File: src/frontend/src/pages/myfeeds.js (hunk 1 above)
Comment (humphd) [reply src/frontend/src/pages/myfeeds.js:157]: We need the stuff that @Grommers00 and @c3ho are doing to get "my" feeds vs. all.

File: src/frontend/src/pages/myfeeds.js [hunk 7]
```diff
@@ -15,94 +34,312 @@ const useStyles = makeStyles((theme) => ({
...
+      fetch(`${telescopeUrl}/feeds`).then((res) => res.json()),
+    ]);
+    const allFeeds = await Promise.all(
+      feedItems.map((item) => fetch(`${telescopeUrl}${item.url}`))
+    ).then((responses) => Promise.all(responses.map((res) => res.json())));
+
+    if (allFeeds.length && user) {
```
Comment (humphd) [inline src/frontend/src/pages/myfeeds.js:160]: This needs to happen at the server.  I know why you're doing it here for testing, but let's not land this.

File: src/frontend/src/pages/myfeeds.js [hunk 8]
```diff
@@ -15,94 +34,312 @@ const useStyles = makeStyles((theme) => ({
...
       const response = await fetch(`${telescopeUrl}/feeds`, {
         method: 'post',
         headers: {
           'Content-Type': 'application/json',
         },
         body: JSON.stringify({
-          author: feedAuthor,
-          url: feedUrl,
+          user: userInfo.id,
+          author: newFeedAuthor,
```
Comment (humphd) [inline src/frontend/src/pages/myfeeds.js:191]: We should probably do something to use `user.name` in the form so the user doesn't have to type it every time.  If they want to change it, they can, but we should set it as a default.

File: src/frontend/src/pages/myfeeds.js [hunk 9]
```diff
@@ -15,94 +34,312 @@ const useStyles = makeStyles((theme) => ({
...
       });
-      console.log(response);
-      return response;
+      await setSubmitStatus(
+        response.ok
+          ? { message: 'Feed added successfully' }
+          : { message: `Error: ${response.status} ${response.statusText}`, isError: true }
+      );
+      if (response.ok) {
+        window.location.reload(false);
```
Comment (humphd) [inline src/frontend/src/pages/myfeeds.js:201]: Why?

File: src/frontend/src/pages/myfeeds.js (hunk 2 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js:22]: Since `valid-url` is already included in our root-level `package.json`, I did not think it would be a good idea to also include it within our frontend-level `package.json`. As such, I added this line to prevent ESLint complaining about its lack of inclusion there.

https://github.com/Seneca-CDOT/telescope/blob/9cba5ed46c166f620c21f19f16baeccc383a44f6/package.json#L94

Comment (humphd) [reply src/frontend/src/pages/myfeeds.js:22]: I wonder if we should just disable this rule in our eslintrc, since we have a few places where it would be good to rely on deps futher up.

Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js:22]: Probably a good idea; should I file an issue?

Comment (humphd) [reply src/frontend/src/pages/myfeeds.js:22]: I'd do it in this, so you can remove that eslint ignore.

File: src/frontend/src/pages/myfeeds.js (hunk 5 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js:108]: Posted an idea here: https://github.com/Seneca-CDOT/telescope/issues/945#issuecomment-609954691

File: src/frontend/src/pages/myfeeds.js (hunk 8 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js:99]: Updated in https://github.com/Seneca-CDOT/telescope/pull/933/commits/2ae7efe11944d53d23e43c897c561cf9948cfedb#diff-341423ac4fcd5b3e25e7dd93d2bae8e1R46: the user's name should populate the author form field by default, now.

File: src/frontend/src/pages/myfeeds.js (hunk 6 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js]: > All of this can/will fail at some point, we need proper error handling. You can chat with @c3ho or @raygervais, who have recently done code like this with error handling.

Although we've recognized the fetching done here is mostly a stop-gap for proper backend logic, I refactored the fetch requests a bit in https://github.com/Seneca-CDOT/telescope/pull/933/commits/2ae7efe11944d53d23e43c897c561cf9948cfedb, wrapping them in try/catch blocks similar to those I've observed @c3ho and @raygervais using in the backend.

> I also think we should probably move the call to get user info to the top level of this page, and already know who you are when you access the content. That's something for #760.

Agreed. For what it's worth, as another temporary measure, I separated the logic for fetching user-info from that of fetching feeds by placing it within a separate useEffects call (https://github.com/Seneca-CDOT/telescope/pull/933/commits/2ae7efe11944d53d23e43c897c561cf9948cfedb#diff-341423ac4fcd5b3e25e7dd93d2bae8e1R35)

File: src/frontend/src/pages/myfeeds.js (hunk 3 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js]: > File a follow-up issue for this, and put the URL to the issue here.

Done: https://github.com/Seneca-CDOT/telescope/issues/946
https://github.com/Seneca-CDOT/telescope/pull/933/commits/2ae7efe11944d53d23e43c897c561cf9948cfedb#diff-c1747a45c828916ed2e3274bdd5df4dcR35

> Also, cc @c3ho and @Grommers00, who are working on backend bits for this.

I'll definitely be in touch with Calvin and James via #946. I'm very curious to learn more about the My Feeds-related backend functionalities they've been working on.

File: src/frontend/src/pages/myfeeds.js (hunk 2 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js]: Done: https://github.com/Seneca-CDOT/telescope/pull/933/commits/2ae7efe11944d53d23e43c897c561cf9948cfedb#diff-e4403a877d80de653400d88d85e4801aR108

Regarding ESLint changes: I also took the liberty of installing the `eslint-plugin-react-hooks` plugin, which I have been finding useful while learning about and working with React Hooks.

Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js]: > Can we break myfeeds.js up into different files? We have a bunch of separate components all being defined in this file, and it's making it hard to read the code.

Good idea, and done in https://github.com/Seneca-CDOT/telescope/pull/933/commits/2ae7efe11944d53d23e43c897c561cf9948cfedb:
- [DeleteFeedDialogButton](https://github.com/Seneca-CDOT/telescope/pull/933/commits/2ae7efe11944d53d23e43c897c561cf9948cfedb#diff-c1747a45c828916ed2e3274bdd5df4dc)
- [HelpPopoverButton](https://github.com/Seneca-CDOT/telescope/pull/933/commits/2ae7efe11944d53d23e43c897c561cf9948cfedb#diff-2b3be8acde3c2c96452d43f50b33ef2a)
- [ExistingFeedList](https://github.com/Seneca-CDOT/telescope/pull/933/commits/2ae7efe11944d53d23e43c897c561cf9948cfedb#diff-2117800b114a1b4cfaa8fb1e262360ea)

File: src/frontend/src/pages/myfeeds.js (hunk 9 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js]: Programmatically reloading the page was a quick and dirty way to reset form fields (`react-material-ui-form-validator` doesn't appear to offer an easy way to do so) while also forcing a rerender. I've now removed these refresh() calls in https://github.com/Seneca-CDOT/telescope/pull/933/commits/2ae7efe11944d53d23e43c897c561cf9948cfedb, having invested some time tonight becoming more comfortable with the `useEffect` hook.

File: src/frontend/src/pages/myfeeds.js (hunk 4 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js]: As per https://github.com/Seneca-CDOT/telescope/pull/933#discussion_r404623978:
> Programmatically reloading the page was a quick and dirty way to reset form fields (`react-material-ui-form-validator` doesn't appear to offer an easy way to do so) while also forcing a rerender. I've now removed these refresh() calls in https://github.com/Seneca-CDOT/telescope/pull/933/commits/2ae7efe11944d53d23e43c897c561cf9948cfedb, having invested some time tonight becoming more comfortable with the `useEffect` hook.

File: src/frontend/src/pages/myfeeds.js (hunk 7 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js]: > We need the stuff that @Grommers00 and @c3ho are doing to get "my" feeds vs. all.
> This needs to happen at the server. I know why you're doing it here for testing, but let's not land this.

Agree 100%. I hope @Grommers00 and @c3ho may assist me with developing web API routes (if none exist already) such that the My Feeds component can make use of their work.

File: src/frontend/src/components/MyFeedsPage/DeleteFeedDialogButton.jsx [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+const useStyles = makeStyles(() => ({
+  button: {
+    padding: '3px 0 3px 0',
+  },
+}));
+
+function DeleteFeedDialogButton({ feed }) {
+  const classes = useStyles();
+  const [open, setOpen] = React.useState(false);
```
Comment (c3ho) [inline src/frontend/src/components/MyFeedsPage/DeleteFeedDialogButton.jsx:23]: Minor nit, but I guess we should be consistent across the whole site with `React.useState` or `React.useEffect` vs doing it in the import `import React, { useState } from 'react'`

File: src/frontend/src/pages/myfeeds.js [hunk 11]
```diff
@@ -37,72 +95,108 @@ export default function MyFeeds() {
...
+      console.log('Error adding feeds', error);
+    }
+  }
+
+  function handleChange(event) {
+    const { name, value } = event.target;
+    if (name === 'author') {
+      setNewFeedAuthor(value);
+    } else {
+      setNewFeedUrl(value);
```
Comment (c3ho) [inline src/frontend/src/pages/myfeeds.js:119]: @humphd I'm surprised this part works, as I thought we couldn't do conditionals with hooks.

File: src/frontend/src/components/MyFeedsPage/DeleteFeedDialogButton.jsx (hunk 10 above)
Comment (Silvyre) [reply src/frontend/src/components/MyFeedsPage/DeleteFeedDialogButton.jsx]: Good spot, fixed in https://github.com/Seneca-CDOT/telescope/pull/933/commits/3a4bfaec30edef643f85006e0ddd078a10c0ecac

File: src/frontend/src/pages/myfeeds.js (hunk 7 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js]: I've supplemented my makeshift `GET /feeds` solution with a fetch to `GET /user/feeds` in https://github.com/Seneca-CDOT/telescope/pull/933/commits/79712f5943c91d678d2e775750abfd8f9762fe88:

https://github.com/Seneca-CDOT/telescope/blob/79712f5943c91d678d2e775750abfd8f9762fe88/src/frontend/src/pages/myfeeds.js#L57-L90

File: src/frontend/src/pages/myfeeds.js (hunk 1 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js]: I've supplemented my makeshift `GET /feeds` solution with a fetch to `GET /user/feeds` in https://github.com/Seneca-CDOT/telescope/pull/933/commits/79712f5943c91d678d2e775750abfd8f9762fe88:
 
https://github.com/Seneca-CDOT/telescope/blob/79712f5943c91d678d2e775750abfd8f9762fe88/src/frontend/src/pages/myfeeds.js#L57-L90
 
The idea is to remove the old, makeshift logic once `GET /user/feeds` is working.

File: src/frontend/src/components/MyFeedsPage/DeleteFeedDialogButton.jsx [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        </DialogActions>
+      </Dialog>
+    </div>
+  );
+}
+
+DeleteFeedDialogButton.propTypes = {
+  feed: PropTypes.object,
+  // 'feed.id': PropTypes.string,
+  // 'feed.url': PropTypes.string,
```
[the remark this answers is not in the document]
Comment (Silvyre) [reply src/frontend/src/components/MyFeedsPage/DeleteFeedDialogButton.jsx]: Nope, good spot; removed in https://github.com/Seneca-CDOT/telescope/pull/933/commits/bd260d87a226f836bc31fd641fc659f57a732b28

**Comment by Silvyre** [comment]
Thanks for your two approving reviews (which I have just rebased out of existence)! Squashing and merging!
